File: prototype/rails/src/test/domain/orders/shipment_details_test.rb
require "test_helper"

module Domain
  module Orders
    class ShipmentDetailsTest < ActiveSupport::TestCase
      test "a carrier and a tracking number are complete" do
        assert_predicate submitted(carrier: "Royal Mail", tracking_number: "RM123456789GB"), :complete?
      end

      test "surrounding space is not part of either field" do
        details = submitted(carrier: "  Royal Mail  ", tracking_number: "  RM123456789GB  ")

        assert_equal "Royal Mail", details.carrier
        assert_equal "RM123456789GB", details.tracking_number
      end

      test "a shipment with no carrier is incomplete" do
        refute_predicate submitted(carrier: " ", tracking_number: "RM123456789GB"), :complete?
      end

      test "a shipment with no tracking number is incomplete" do
        refute_predicate submitted(carrier: "Royal Mail", tracking_number: ""), :complete?
      end

      test "a missing field is incomplete rather than an error" do
        refute_predicate submitted(carrier: nil, tracking_number: nil), :complete?
      end

      private

      def submitted(carrier:, tracking_number:)
        ShipmentDetails.from_input(carrier: carrier, tracking_number: tracking_number)
      end
    end
  end
end

File: prototype/node/src/app/core/orders/shipment-details.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { parseShipmentDetails } from './shipment-details.ts'

test('a carrier and a tracking number parse into a shipment', () => {
  const parsed = parseShipmentDetails({ carrier: 'Royal Mail', trackingNumber: 'RM123456789GB' })

  assert.equal(parsed.ok, true)
  if (!parsed.ok) return
  assert.equal(parsed.value.carrier, 'Royal Mail')
  assert.equal(parsed.value.trackingNumber, 'RM123456789GB')
})

test('surrounding space is not part of either field', () => {
  const parsed = parseShipmentDetails({ carrier: '  Royal Mail  ', trackingNumber: '  RM123456789GB  ' })

  assert.equal(parsed.ok, true)
  if (!parsed.ok) return
  assert.equal(parsed.value.carrier, 'Royal Mail')
  assert.equal(parsed.value.trackingNumber, 'RM123456789GB')
})

test('a shipment with no carrier is refused', () => {
  const parsed = parseShipmentDetails({ carrier: ' ', trackingNumber: 'RM123456789GB' })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, { carrier: 'Enter the carrier.' })
})

test('a shipment with no tracking number is refused', () => {
  const parsed = parseShipmentDetails({ carrier: 'Royal Mail', trackingNumber: '' })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, { trackingNumber: 'Enter the tracking number.' })
})

test('a missing field is an error rather than a blank shipment', () => {
  const parsed = parseShipmentDetails({ carrier: null, trackingNumber: null })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, {
    carrier: 'Enter the carrier.',
    trackingNumber: 'Enter the tracking number.',
  })
})
